Add support for all amsmath symbols and test/reference.html for validation

arnog opened the pull request:
This test files includes the symbols and commands from the AMS short-math-guide (ftp://ftp.ams.org/pub/tex/doc/amsmath/short-math-guide.pdf) as they are displayed by MathQuill and LaTeX (via codecogs) for reference.

eoghanmcilwaine committed:
[LRN] Allow direct entry of $ (dollar sign)

By removing the binding that creates a text block
when $ is pressed.

laughinghan committed:
API to enable "fast clicks" on touch devices

Argghhhhh. Want touch swipes to scroll the page, even if they start in
MathQuill (so that it's possible to scroll an expression list filled to
the brim with nothing but MathQuills), but touch taps to place the
cursor. Now, with mouse events, if all you care about is a logical
"click" on the current platform, you just listen for the click event and
the browser will take care of whether a particular sequence of
mousedown-mousemoves-mouseup is a click or not. With touch events,
there's no equivalent logical "tap" event, and while legacy mouse events
do fire if you don't preventDefault() on touchstart, Safari on iOS waits
a glacial ~360ms [1] after a tap in case you're actually doing the
double-tap gesture to zoom in.

[1]: Videos: http://developer.telerik.com/featured/300-ms-click-delay-ios-8/

It get's better. Not only are the legacy mouse events unhelpful, they're
misleading, because there is no way to directly tell them apart from
real mouse events and no way to get them not to fire except calling
preventDefault(), which breaks touch-swipe-to-scroll.

As a result, across the Web there are dozens of blog posts about, and
shitty JS libraries for, "fast clicks", that all disagree on what counts
as a "tap", and also when to ignore a mouse event because it's probably
a legacy one 300ms after a touch event that's already been dealt with.
By far the 2 most-cited "fast click" solutions are Google Developer's
"Fast Buttons" article [2] and FT Labs' FastClick library [3]; only the
latter ignores taps that are to stop scrolling after a fling, taps where
touchend is >700ms after touchstart, or taps with more than one finger,
but only the former appears to put a time and distance limit on mouse
events to ignore (FastClick just ignores the next one).

[2]: https://developers.google.com/mobile/articles/fast_buttons
[3]: https://github.com/ftlabs/fastclick

MathQuill's sponsor, Desmos, has a custom touch events library whose
definition of "tap", and whose strategy for ignoring delayed mouse
events, is different from either of the above. Clearly, this stuff needs
to be decided by the parent application using MathQuill.

So, with heart-wrenching despair, here are 2 new API calls:

- .fastClick(target, clientX, clientY) takes in the touch event info
  and places the cursor like there was a click there

- .ignoreNextMousedown(fn) takes a function that is called on subsequent
  mousedown events, which MathQuill will ignore if true is returned.
  MathQuill stops calling the function once it returns false.
  Meant to be called on touchend before every expected legacy mousedown,
  regardless of whether there was a logical "tap".

--

The test case has a trivial notion of logical "tap", which is if no
touchmoves happen between the touchstart and touchend. Note that it
calls .ignoreNextMousedown() after every touchend, or else sometimes if
you wiggle your finger a little but not much, the logical "tap" won't
happen but Safari on iOS will still fire a legacy mousedown event.

Test case also gained -webkit-tap-highlight-color to to hide the gray
tap highlight box.

laughinghan committed:
Upgrade to Mocha 2.4 to fix iffy "stealing back" focus test

Recommend -w to view diff.

That test requires the document to be focused for the browser to exhibit
the "stealing back" focus behavior, so we had a check to skip the test
if the browser isn't focused.

Unfortunately that check was useless a large fraction of the time due to
a race condition: if you Refresh unit.html, while tests are all being
defined but none are being run the check happens, and then while the
whole test suite is actually running (slow), you switch to another
browser tab, then when the test actually runs the document is unfocused
and the test fails.

(Note that this only happens if you have the Developer Tools/Web
 Inspector closed; when they're open and focused the document will be
 unfocused anyway, even when you Refresh. So to reproduce that above
 behavior make sure Dev Tools aren't open on unit.html when you Refresh.)

laughinghan committed:
Fix focusBlur test suite when document not focused

laughinghan committed:
Makefile nice err msg if no Node or installed wrong

In response to this today:
https://mathquill.slack.com/archives/mathquill/p1458182072000301

Holy *shitsnacks* I forgot how much Make sucks. I literally can't quote
my argument to the `ifneq` conditional.

And of course, I'm writing this to workaround the fucking Debian package
repo maintainers...

laughinghan committed:
Feature: integrals' from/to now mandatory, prettier

See the GitHub comment on this commit for screenshots. (Also copied below.)

Like #258 but for integrals rather than sums and products, integrals now
come with from and to blocks as an inextricable part of them.

Note that this means that integrals with only a subscript, e.g.
Stoke's theorem `\int_{\partal\Omega} \omega = \int_\Omega d\omega`,
can't be typed anymore, but #258 made an analogous sacrifice.

I originally intended this to be two PRs, one to fix how integrals look
and one to make the from/to blocks mandatory, but was stymied by the
vertical alignment of the superscript block (see the second test case).
The only way I could figure out to get that to work was .mq-sup-inner,
which doesn't exist in normal sup/sub blocks, I'm only able to add it by
making from/to blocks mandatory and inserting a span.mq-sup-inner into
the HTML template (luckily mandatory from/to blocks are what Desmos
needs anyway).

One nice thing: supsub no longer has to autodetect whether it's
currently serving as the from/to of an integral, we can get rid of all
the .mq-limit stuff AND the SupSub::respace() stuff, and get rid of it
from the redrawing/reflowing test.

Speaking of tests, there wasn't enough space for the spherical integral
in the rendered math column, so I made it wider, half instead of a third
of the page.

----
Before: <img width="238" alt="before" src="https://cloud.githubusercontent.com/assets/225809/14314220/d69bd5fc-fbaa-11e5-8487-2c79eef5bb1d.png">
After: <img width="241" alt="after" src="https://cloud.githubusercontent.com/assets/225809/14314221/d69e7118-fbaa-11e5-8954-551a5c06d495.png">
KaTeX: <img width="354" alt="katex" src="https://cloud.githubusercontent.com/assets/225809/14314365/9b258e2c-fbab-11e5-8594-d2e7617b4185.png">
[MathJax](http://mathb.in/55065): <img width="433" alt="mathjax" src="https://cloud.githubusercontent.com/assets/225809/14314364/9b2348b0-fbab-11e5-80eb-31a58c4910d1.png">
LaTeXiT: <img width="354" alt="katex" src="https://cloud.githubusercontent.com/assets/225809/14314355/92c3d07c-fbab-11e5-9d73-296a599e7a3a.png">

laughinghan committed:
Add unit test for integrals' mandatory from/to blocks

And test interaction with sumStartsWithNEquals

laughinghan committed:
Fix auto-commands eating letters from operator names

Should have lower "precedence".

Auto-commands used to be the first thing that Letter::createLeftOf()
checked for. Now instead it first waits for the dust to settle on
auto-operator names, then seeks nearby letters but stops at the edge of
an operator name. Also note that before, this letter itself had not been
inserted yet; now it has so the seeking and the fragment that's removed
both start from the this letter.

stufflebear committed:
Moved usage information from README to docs folder

stufflebear committed:
rename to docs/index.md

stufflebear committed:
Update README with more information

Includes basic summary of what MathQuill is, demo information, partners, and CoC

stufflebear committed:
Create Code of Conduct

Original Issue: https://github.com/mathquill/mathquill/issues/564

stufflebear committed:
Rename CODE_OF_CONDUCT.md to Code_of_Conduct.md

stufflebear committed:
Update Code of Conduct link

From Google Docs draft to Read the Docs link

laughinghan committed:
Fix 'test florin spacing' to clean up after itself

laughinghan committed:
Fix sin^2(x), and sin^2 x as cursor moves through #visual

See GitHub comment for screenshots/GIFs (source copied below).

Before (notice space before paren):
<img width="176" alt="static_before" src="https://cloud.githubusercontent.com/assets/225809/14372074/280414d4-fcf2-11e5-8652-fc44eb8eb3cb.png">
![cursor_after](https://cloud.githubusercontent.com/assets/225809/14372075/2806fe56-fcf2-11e5-9779-a4d49f573cb1.gif)

After:
<img width="183" alt="static_after" src="https://cloud.githubusercontent.com/assets/225809/14372073/280404e4-fcf2-11e5-94ff-ec9fbc59f41d.png">
![cursor_before](https://cloud.githubusercontent.com/assets/225809/14372072/28037ec0-fcf2-11e5-87a2-9d86b1c40677.gif)

stufflebear committed:
Added Contrbuting Guide and Refactored some documentation

stufflebear committed:
Factored remaining docs out of index

stufflebear committed:
Updated APi_Methods, Config, and README

Updated content and formatting.

stufflebear committed:
Filled in links and some minor fixes

laughinghan committed:
Merge branch 'master' into fix.sin-squared

 Conflicts:
	test/unit/css.test.js
	#607 also appended a test

arnog committed:
Added test/reference.html

The reference.html file includes the symbols & commands from the AMS
short-math-guide as they are displayed by MathQuill and LaTeX (via
codecogs) for comparison.

arnog commented:
Still a work in progress, there are some unfinished sections to cover, but feedback welcome.

arnog committed:
Complete coverage of short-math-guide

Section 3.12, 3.13, 3.15, 3.17, 3.18, 4, 5, 6, 7, 8

arnog commented:
OK, I've pushed the rest of the changes. The file should now include coverage for all the relevant sections of short-math-guide. For some reason the commit is not showing yet in GitHub, but any day now...

laughinghan commented on test/reference.html:
```diff
@@ -0,0 +1,946 @@
+    $('#table').html(t);
+    $('.mathquill-static-math').each(function() { 
+        MQ.StaticMath(this);
+        if (this.textContent === '$$\u200B') {
```
this is very dependent on implementation details, I'd recommend using the API, like `MQ(this).latex() === ''`

arnog replied:
Ah, yes, much better. Thanks.

laughinghan commented on test/reference.html:
```diff
@@ -0,0 +1,946 @@
+}
+
+img {
+    opacity: .5;
```
why is this?

arnog replied:
The opacity? I was trying to find some way to make the codecogs rendering less prominent. I've removed the opacity trick and just change the background color now.

laughinghan commented:
This is awesome! Two moderate concerns:
- The page is very wide, much wider than the Short Math Guide PDF itself even blown up to an equivalent font size, can we get at least the minimum width to be less wide? (If you like wide that's fine, but it should be fluid to smaller widths too within reason.) Width of a PDF at the same font size is usually my rule of thumb
- Every page load of this sends >600 requests to CodeCogs, right? Is that kosher? It's not clear whether or not that conforms to their Usage Policy, which is clearly written with usage on public websites in mind, not local test pages: http://www.codecogs.com/latex/usage.php
    + one other popular alternative is the Google Charts API but that's deprecated apparently? https://github.com/adam-p/markdown-here/issues/261
    + there's also QuickLatex? http://quicklatex.com/
    + I dunno what the right answer here is; other than Google I'd be similarly concerned with spamming requests to any such 3rd-party service
    + other alternatives include rendering locally (MathJax can output SVG, right? Or is there too much that's not supported by MathJax?), or saving all the images (seems unfortunate to bloat our repo, but maybe they're tiny)

laughinghan commented:
weirder idea: pre-render all the images but instead of keeping them in our repo, upload them to the git asset storage used for storing images in comments and issues ![image](https://cloud.githubusercontent.com/assets/225809/14418957/287510da-ff77-11e5-994c-f0097b18a834.png), and use that link

arnog committed:
Added MathJax support

Tweaked the layout, fixed some typos, added some notes, and added
support for MathJax in addition to codecogs as a reference to compare
against.

laughinghan commented on test/reference.html:
```diff
@@ -44,48 +58,72 @@
-
-img {
-    opacity: .5;
+    display: inline-block;
```
images are replaced elements, aren't they already effectively `inline-block`?

arnog replied:
Yes, I was trying to get the alignment to work right. Fixed it now, don't need this anymore.

arnog commented:
I've adjusted the width and the layout. Looking at it printed, it looks pretty good to me, but let me know what you think.

The codecogs TOU say "you can generation up to 3,000* equations a day for testing and non-commercial use". That would seem to cover this local test page, but I get your point.

I've added support for MathJax and commented out the codecogs rendering. The MathJax rendering is pretty good (in fact better than the codecogs one, as it weirdly cuts off the top pixel, at least in SVG), The only notable differences I could spot are no support for \slash, \thickspace, \medspace, \displaylimits, unmatched \begin{?matrix} \end{?matrix} and optional alignment parameters in \cfrac.

arnog committed:
Comment out codecogs

laughinghan commented:
Definitely looking even better, though still getting some width weirdness:
![image](https://cloud.githubusercontent.com/assets/225809/14437671/35f5136c-ffd7-11e5-8b1b-607ed28df31e.png)

laughinghan commented on test/reference.html:
```diff
@@ -907,14 +990,28 @@
                     t += '<td class="mathquill-static-math">' + entries[j + 1] + '</td>';
-                    t += '<td><img src="' + 'http://latex.codecogs.com/svg.latex?' + encodeURI(entries[j + 1]) + '"></td>';
+                    t += '<td class="codecogs"><img src="' + 'http://latex.codecogs.com/svg.latex?' + encodeURI(entries[j + 1]) + '"></td>';
+                    t += '<td class="mathjax">$$' + entries[j + 1] + '$$</td>';
```
does this mean we're relying on the MathJax feature where it searches through every text node on the page for `$$`? Since we're generating the elements it might be a lot faster to call the MathJax API directly on our known elements (like by querying for all elements with the `.mathjax` class and MathJax-ifying them or something)

arnog replied:
I've switched to deferring the typesetting and making it explicit, but I don't see any difference in speed. Not sure why you think it would be faster. Scanning all the nodes to find the $$ is negligible compared to the cost of manipulating the DOM and redoing the layout.

laughinghan replied:
Damn, you're totally right

Han

On Mon, Apr 11, 2016 at 12:16 PM, Arno Gourdol <notifications@github.com>
wrote:

> In test/reference.html
> <https://github.com/mathquill/mathquill/pull/615#discussion_r59263817>:
>
> >              t += '<table>';
> >              while (j < entries.length) {
> >                  if (entries[j] === sections[i]) {
> >                      // entry matches the current section
> >                      t += '<tr>';
> >                      t += '<td class="mathquill-static-math">' + entries[j + 1] + '</td>';
> > -                    t += '<td><img src="' + 'http://latex.codecogs.com/svg.latex?' + encodeURI(entries[j + 1]) + '"></td>';
> > +                    t += '<td class="codecogs"><img src="' + 'http://latex.codecogs.com/svg.latex?' + encodeURI(entries[j + 1]) + '"></td>';
> > +                    t += '<td class="mathjax">$$' + entries[j + 1] + '$$</td>';
>
> I've switched to deferring the typesetting and making it explicit, but I
> don't see any difference in speed. Not sure why you think it would be
> faster. Scanning all the nodes to find the $$ is negligible compared to the
> cost of manipulating the DOM and redoing the layout.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mathquill/mathquill/pull/615/files/10ce7424d9558525a74ad1ef00fca9c61248a8d2..6810806ca08bcb5f8cc1d3b3ce9e29465df38b93#r59263817>
>

arnog committed:
Narrower page

Adjustments to display more narrowly.
Also, better way to recognize that Mathquilification failed.

arnog committed:
Defer the MathJax typesetting

Don't automatically typeset MathJax content on startup. Do it explicitly
after all the content has been generated.

arnog committed:
Added non-standard commands

Added a list as an Appendix of commands that are not part of amsmath but
are supported by MathQuill, including some MathQuill synonyms, variants
for Greek letters and MathQuill specific commands.

arnog committed:
Added \lbrack, \rbrack, \lcurlybrace and \rcurlybrace

arnog committed:
Added non-standard /ang command

arnog committed:
Add support for the set of standard amsmath LaTeX commands

These are mostly additive changes (i.e. adding a definition for a
command that did not exist before), but there are a few notable
exceptions listed below.

A remaining issue is that there is a larger set of delimiters recognized
as "Bracket". However, delimiters that are "Bracket" can't seem to
function independently. That is, "\lceil x \rceil" does not render, but
"\left\lceil x \right\rceil" does.​

\deg was aliased to \degree *AND* defined as a built-in function. \deg
*is* a standard amsmath commands, therefore dropped the alias

\ast was aliased to \star, but they are different standard amsmath
commands

\asymp was aliased to \approx, but they are different standard amsmath
commands

\diamond, \bigtriangleup, \ominus, \uplus, \bigtriangledown, \sqcap,
\triangleleft, \sqcup, \triangleright, \odot, \circledot, \bigcirc,
\dagger, \ddager, \wr and \amalg were defined as VanillaSymbol, but they
are BinaryOperators.

\models, \prec, \succ, \preceq, \succeq, \simeq, \mid, \ll, \gg,
\parallel, \nparallel, \bowtie, \sqsubset, \sqsupset, \smile,
sqsubseteq, \sqsupseteq, \doteq, \frown, \vdash, \dashv, \nless, \ngtr
were defined as VanillaSymbols, but they are BinaryOperators (comparison
operators)

arrows and harpoons were defined as VanillSymbol but they are
BinaryOperatos
\oint, \bigcap, \bigcup, \bigsqcup, \bigvee, \bigwedge, \bigodot,
\bigotimes, \bitoplus, \biguplus were defined as VanillaSymbol, but they
are variable size operators (SumationNotation)

\implies was aliased to \Rightarrow, but it should be to \Longrightarrow

\iff was aliased to \Leftrightarrow, but it should be to
\Longleftrightarrow

\oslash was aliased to \varnothing, but there is a standard amsmath
command, therefore dropped this alias

\Upsilon was mapped to &upsih; U+03d2 (GREEK UPSILON WITH HOOK SYMBOL).
It should be mapped to U+03a5 (GREEK CAPITAL LETTER UPSILON). \upsih and
\Upsih now maps to &upsih;

\wedge was mapped to \land. They are distinct commands.

\vee was mapped to \lor. They are distinct commands.

arnog commented:
Added support for the amsmath commands. 

There is one issue remaining with delimiters: they only function with \left + \right commands. Some delimiters were previously defined as non-Bracket (i.e. \lfloor) and were displayed fine standalone. Now, they only display if after a \left. This doesn't affect text entry (which will generate the \left automatically), but it does affect Latex conversion.

Otherwise, things look good. See before and after PDF.

[reference-before.pdf](https://github.com/mathquill/mathquill/files/216164/reference-before.pdf)
[reference.after.pdf](https://github.com/mathquill/mathquill/files/216165/reference.after.pdf)

arnog commented:
Also worth noting, Symbola is missing a few codepoints: \gvertneqq, \lvertneqq, \ngeqq, \ngeqslant, \nleqq, \nleqslant. Perhaps those could be simulated with a \not + appropriate command?

arnog committed:
Add \class command

arnog committed:
Add more text commands in reference.html

arnog committed:
Revert extra delimiters

Give up for now on having delimiters work both with \left and without.
Back to the level of support that was there before or better (more
delimiters supported now). Will need a more general solution, probably
involving some Bracket surgery...

laughinghan commented:
Man, this is amazing, a whole mass of new symbols!

There's still some funkiness, but obviously this PR doesn't have to be perfect and fix everything to be merged, it just has to be an improvement. The main thing I'm looking for that would block merging is regressions, and anything else worrisome like the ">600 calls to CodeCogs"-type miscellanea.

- [ ] How can we verify that there are no regressions? I'm thinking we dump the old and new `LatexCmds` to JSON and use a script to compare and make sure the Unicode characters and symbol class (`BinaryOperator` etc) of anything old is the same in the new, or any discrepancies are documented by 9398c4e4baff2e31b4a30528eeb0bdd1d58251ea
- [ ] Can you think of anything else worth double-checking?

----

Stuff I noticed but might not block merging:
- We should comment-out the mention of CodeCogs if we're gonna comment out its usage
- Looks like you made integrals into having `\limits`, so they're displayed like `\sum`, I think that'll conflict with #603
    + Are the new integral/sum/etc signs too big?
- 3.16 Accents&mdash;these are mostly so broken they probably shouldn't be added at all

Stuff I noticed but definitely doesn't block merging:
- MathJax shows up twice for a while? A little Googling suggests maybe it's the PreviewHTML? Can we disable that?
- A lot of the relation symbols like `\approxeq` are `VanillaSymbol`s, they should be `BinaryOperator`s like `=`
- `\vec\imath` and `\hat\jmath` look terrible because of the `text-align: right`

laughinghan commented on test/reference.html:
```diff
@@ -1050,6 +1053,13 @@
+    // We've deferred the typeseting on startup. Do it explicitly now on each
+    // elements with a "mathjax" class.
+    $('.mathjax').each(function() { 
+        MathJax.Hub.Queue(["Typeset", MathJax.Hub, this]);
```
if I comment this out, MathJax still auto-renders&mdash;I don't think this commit did anything

laughinghan commented on src/commands/math/advancedSymbols.js:
```diff
@@ -2,43 +2,202 @@
+'bigcirc:25ef', 'bullet:2219', 'ddagger:2021', 'wr:2240', 'amalg:2a3f',
+
+'cdot:22c5', 'circ:2218', 'div:00f7', 'pm:00b1', /* 'times:00d7', */ 'cap:2229',
+'cup:222a', 'setminus:2216', 'land:2227', 'lor:2228', 'wedge:2227',
```
how are `\land` and `\wedge` distinct? Seems like they're the same class and the same character?

arnog replied:
They are indeed synonyms. \land, \lor and \lnot are more commonly used, while \wedge, \vee and \neg are used by lattice theory specialists.

laughinghan commented on src/commands/math/basicSymbols.js:
```diff
@@ -491,3 +422,88 @@ LatexCmds['÷'] = LatexCmds.div = LatexCmds.divide = LatexCmds.divides =
+// MathQuill non-standard commands
+LatexCmds.upsih = //W3C/Unicode "upsilon with hook"
+LatexCmds.Upsih = //'cos it makes sense to me
+   bind(Symbol,'\\Upsilon ','<var class="mq-nonSymbola">&upsih;</var>');
```
This probably shouldn't be upright, not `<var>`, right? Just like capital Upsilon (capital Greek letters are customarily upright rather than italicized in math for some reason)

arnog replied:
Good point. This was the original code: ```bind(Symbol,'\\Upsilon ','<var style="font-family: serif">&upsih;</var>'); ``` but that was probably incorrect as well.

Come to think of it, this probably shouldn't have a control sequence of \Upsilon anymore, since this is now distinct from regular Upsilon. Probably make sense to map it to \upsi, no?

laughinghan replied:
> Come to think of it, this probably shouldn't have a control sequence of \Upsilon anymore, since this is now distinct from regular Upsilon. Probably make sense to map it to \upsi, no?
 
Oh shoot, I was only thinking about backcompat when importing LaTeX, I forgot about interop when exporting. There's no real AMS-LaTeX command for upsilon with a hook, right? The only way for this to interop with MathJax is probably `\unicode{}`, and no way with KaTeX (except Unicode in the source), right? I'm thinking we should just comment this out and note this as a breaking change.

arnog replied:
OK, but ```\upsih``` is not the only command that doesn't interop with AMS-LaTeX or MathJax. For example ```\parallelogram```.

laughinghan commented on src/commands/math/basicSymbols.js:
```diff
@@ -491,3 +422,88 @@ LatexCmds['÷'] = LatexCmds.div = LatexCmds.divide = LatexCmds.divides =
+}
+
+LatexCmds['#'] = bind(VanillaSymbol, '\\# ', '#');
+LatexCmds['_'] = bind(VanillaSymbol, '\\_ ', '_');
```
I don't think we can successfully parse `\#` and `\_`...

arnog replied:
```\#``` seems to work (trying it out in demo.htm) but you're right for \_. I'll comment it out.

laughinghan commented on src/commands/math/basicSymbols.js:
```diff
@@ -2,6 +2,35 @@
+                    - Inequality
+                    - Equality
+*/
+
```
💯

laughinghan commented on src/commands/math/basicSymbols.js:
```diff
@@ -491,3 +423,90 @@ LatexCmds['÷'] = LatexCmds.div = LatexCmds.divide = LatexCmds.divides =
+for (i = 0; i < BASIC_VANILLA_SYMBOLS.length; i++) {
+    m = BASIC_VANILLA_SYMBOLS[i].match(/([a-zA-Z]+)\/?([a-zA-Z]*):(\w+)/);
+    LatexCmds[m[1]] = bind(VanillaSymbol, '\\' + (m[2] && m[2].length > 0 ? m[2] : m[1]) + ' ', '&#x' + m[3] +';');
+}
```
I'm not sure how often the code you're replacing did this, but this would be the perfect spot to support Unicode when pasting in/parsing LaTeX, right? Either by find-replacing all the code points to start with `\u` escape sequences and using the one character, or I guess `String.fromCharCode()` or something

laughinghan commented:
This really was fantastic work, @arnog!

One more note: I think the ideal solution we'd want to eventually have is to have one file that generates both the `LatexCmds` in source and the documentation/visual test page. Any thoughts on that?

arnog committed:
Faster MathJax loading in reference.html

Switched to using a custom configuration of MathJax and the <script> tag
instead of inline $$. This results in much faster rendering by MathJax.
After correctly disabling initial typesetting I tried doing a manual
typesetting on startup but this resulted in **much** slower rendering
(because each queued typesetting operation results in a
relayout/repaint, which is very, very slow).

arnog commented:
I like the idea of comparing a JSON dump of LatexCmds (and CharCmds?). I'll look into that.

Integrals (like ```\int```, ```\oint```, etc...) and **big operators** like ```\bigvee``` are of class **SummationNotation**. Should they all be **VanillaSymbols** for now until PR#603 is merged? Seems like they look better for now as a **SummationNotation** until something better comes along with PR#603, no? And are you suggesting we only keep ```\int```, ```\sum```, ```\prod``` and ```\coprod``` as SummationNotation?

Yes, the big operators do look a little big. That should probably be adjusted. Anything you have in PR#603 for this?

I agree about the accents, I'll turn them off for now. It's unfortunate that any expression that uses them will now fail rather than displaying something. It would probably be useful to have a lenient option that would still display as much as possible an expression with Latex it cannot parse fully (like MathJax does).

I've optimized the MathJax rendering to avoid the double render and to generally speed it up.

Good catch on the relation symbols. They should indeed be **BinaryOperator**s.

arnog committed:
Addressing first round of feedback

- Math binary relations and negated binary relations (i.e. \nless,
\less, etc..) BinaryOperators instead of VanillaSymbols
- Commented out accent support since they don't display correctly yet
- Commented out \_ since it doesn't parse correctly yet
- Made \upsih a <span> rather than a <var>
- Made the big operators (\bigcap, \bigvee, etc...) VanillaSymbol rather
than SummationNotation since it conflicts with the upcoming PR#603

arnog committed:
In reference.html, small tweak

Rearranging of appendices and reintroduction of red background for
MathJax content

arnog committed:
Improved backward compatibility

- Made some commands (e.g. /bigvee) Symbol instead of SummationNotation.
- re-introduced \π
- made arrows binary operators

arnog commented:
I believe I've addressed all the previous comments. Here is the generated before/after comparison of LatexCmds. Let me know what you think and if there's anything else missing before merging to master.

[collation.pdf](https://github.com/mathquill/mathquill/files/221859/collation.pdf)

laughinghan commented:
The `\upsih` thing made me remember, besides LaTeX import backcompat, the other thing to check for is LaTeX output interop. I expect this not to be a big problem, since these are all from the Short Math Guide to LaTeX, right? Still, how can we verify we haven't added anything that isn't real LaTeX or can't be parsed by MathJax or KaTeX? (Or if we have, we should comment them or something.)

Thank you so much for the new comparison! Unfortunately even just the changed section has almost 300 commands (15 or so per page for 19 pages and change), which is just too many to individually look at. I'm looking for outliers, things that might be regressions or interop problems like the `\deg` conflict with `amsmath` or `\ast` vs `\star`, as noted in your commit message in 9398c4e4baff2e31b4a30528eeb0bdd1d58251ea&mdash;that commit message was perfect, actually, summarizing code changes affecting hundreds of commands in like a dozen items. Can we group these by what the change was, like grouping all the symbols who only change was from `VanillaSymbol` to `BinaryOperator` into a section, etc?

Could you also provide the script you're using to generate these, like in a comment or Gist or something? Also if it would make sense, the script you used to generate these commands from the KaTeX `symbols.js`, if that's still where this is all from? How much manual editing did you do afterwards?

I really can't express enough how impressed I am, by the way. Can I get you cookies or something when this is merged?

arnog commented:
The file ```reference.html``` should be a good indicator regarding the interop of the output since we compare it to MathJax (and codecogs if we enable it). I could add the MathJax rendering for the added commands in the collation document (probably not worth doing for the existing ones, since there are already dozens that are non-standard). Thoughts?

I could try to group the changes in categories, but that will still leave lots of changes to review. I'm not sure there's a way around that. That's how I arrived at the commit message listing the differences.

I didn't use a script to get the symbols from KaTeX, just a RegEx in a text editor. AFAIR I didn't change any of the values afterwards. I'm attaching the source for the collation document to this comment.
[collation.html.txt](https://github.com/mathquill/mathquill/files/222268/collation.html.txt)

laughinghan commented on src/commands/math/advancedSymbols.js:
```diff
@@ -2,43 +2,202 @@
+
+LatexCmds.iff = LatexCmds.Longleftrightarrow;
+LatexCmds.implies = LatexCmds.Longrightarrow;
+LatexCmds.enspace = LatexCmds['\\;'];
```
`LatexCmds['\\;']` doesn't appear to exist, nor `LatexCmds[';']`; removing

laughinghan committed:
Delete \enspace alias to nonexistent \; #cleanup

https://github.com/mathquill/mathquill/pull/615#discussion_r59976804

laughinghan commented:
> I could add the MathJax rendering for the added commands in the collation document (probably not worth doing for the existing ones, since there are already dozens that are non-standard). Thoughts?

I think I was unclear. I don't care about interop of nonstandard aliases like `\epsiv` or `\xists`, because while MathQuill accepts those when importing LaTeX, MathQuill never exports those; when exporting, those are normalized to e.g. `\varepsilon` and `\exists` respectively, which are valid LaTeX commands that correctly interop with MathJax and KaTeX.

I _am_ worried about e.g. `\parallelogram`, because MathQuill _does_ export that, and it doesn't interop with MathJax and KaTeX, which I guess don't support the `gensymb` package, only `amsmath`. Now, `\parallelogram` in particular wasn't added or worsened by this PR, so we can leave it be, but I want to ensure we don't accidentally add non-interoperable commands, if we do at all I want it to be deliberate. For example, I notice that `\sigmav` used to be an alias for, and normalized to, `\varsigma`, but that's broken by this PR: on this PR, if you type <code>\sigmav </code>, MathQuill will output the non-interoperable `\sigmav`, when it should output an interoperable `\varsigma`. A totally reasonable oversight, but a problem we should fix before merging.

On the other hand, `&upsih;` (`ϒ`, U+03D2) seems to have no exact AMS-LaTeX equivalent, since as you pointed out, `\Upsilon` should in fact be `Υ`, U+03A5. That's a candidate for deliberately outputting a non-interoperable control sequence, though I'm leaning towards leaving it as an alias, since I think reducing our interop gap is more important than maintaining this particular, rather subtle distinction.

> I could try to group the changes in categories, but that will still leave lots of changes to review. I'm not sure there's a way around that. That's how I arrived at the commit message listing the differences.

That commit message is exactly what we need, "lots of changes" being like a dozen groups of changes to review, rather than hundreds in no particular order or organization. Thank you for the source to the collation document, I've modified it so the "Commands changed" section has 9 subsections (and re-generated the dump of the original commands to preserve a little more information): https://gist.github.com/laughinghan/80a399fcd492746d9d29522d180d8fb1

To run it, add `MQ.LatexCommands = LatexCmds;` to `src/publicapi.js` (just like yours required) and [delete `\enspace`](https://github.com/mathquill/mathquill/pull/615#discussion_r59976804) ([patch that does both of these things](https://gist.github.com/laughinghan/e5da7e33194b3900da3c69ec6344d8a6)).

Result: [MathQuill Collation 0bc270f.pdf](https://github.com/mathquill/mathquill/files/222582/MathQuill_Collation_0bc270f.pdf)

It only lists 196 changed commands rather than close to 300 as in yours. I believe this is because while yours treated decimal and hex numeric character entity references as equivalent, yours didn't consider e.g. `&#2229;` and `&cap;` equivalent even though they're parsed as the same Unicode character, which I sidestepped by comparing Unicode characters directly.

By the by, the way I noticed the `\sigmav` problem mentioned above was when I was looking at the "Changes to aliases" subsection:
![image](https://cloud.githubusercontent.com/assets/225809/14585915/df4ea488-043b-11e6-9fc5-2c8e2ee22c46.png)

laughinghan committed:
Add back space to control sequences for cumulative operators

Gets rid of \sum, \prod, \coprod, \int, \oint and all the \bigcap,
\bigcup, etc \big* commands from the "Commands removed" section of the
"collation" test page:
  https://gist.github.com/laughinghan/80a399fcd492746d9d29522d180d8fb1

Before: https://github.com/mathquill/mathquill/files/222582/MathQuill_Collation_0bc270f.pdf
After: https://github.com/mathquill/mathquill/files/222592/MathQuill_Collation_after_cleaning_up_cumulative_operators.pdf

laughinghan committed:
Add space to control sequences to delimiter symbols

Gets rid of \lfloor, \rfloor, \lceil, \rceil, \lbrace, \rbrace from the
"Commands removed" section of "collation" test page:
  https://gist.github.com/laughinghan/80a399fcd492746d9d29522d180d8fb1

Before: https://github.com/mathquill/mathquill/files/222592/MathQuill_Collation_after_cleaning_up_cumulative_operators.pdf
After: https://github.com/mathquill/mathquill/files/222593/MathQuill_Collation_after_cleaning_up_delimiter_symbols.pdf

laughinghan committed:
Normalize \lbrack and \rbrack to [ and ] once again

Gets rid of "[" and "]" from the "Commands removed" section of the
"collation" test page:
  https://gist.github.com/laughinghan/80a399fcd492746d9d29522d180d8fb1

Before: https://github.com/mathquill/mathquill/files/222593/MathQuill_Collation_after_cleaning_up_delimiter_symbols.pdf
After: https://github.com/mathquill/mathquill/files/222595/MathQuill_Collation_after_normalizing_lbrack_rbrack.pdf

laughinghan commented:
https://github.com/arnog/mathquill/pull/1 suggests some changes to address things brought up by the new "collation" document, reducing the "Commands removed" section from this:
![image](https://cloud.githubusercontent.com/assets/225809/14586451/515d0022-044d-11e6-8b4d-f7d89e3ea229.png)

To this:
![image](https://cloud.githubusercontent.com/assets/225809/14586458/74dc7776-044d-11e6-8a52-107d2fafe8ae.png)

I'm not sure what to do about the remaining removed command, which appears to be because `\vert` used to alias to the normal ASCII pipe character `|`, and you changed it to map to the fancy Unicode character U+2223. KaTeX appears to render both the normal pipe `|` and `\vert` as U+2223, whereas MathJax renders them both as the normal ASCII pipe. They do seem to agree that normal pipe `|` and `\vert` should be aliases of each other.

laughinghan commented:
Looks like there's some cleanup work left, but the end is in sight and we're set up so that when we're ready to merge, we'll know it with confidence!

arnog committed:
Merge remote-tracking branch 'refs/remotes/mathquill/master'

arnog committed:
Merge pull request #1 from laughinghan/all-amsmath-symbols

Some fixes to reduce divergence from mainline

arnog committed:
Merge branch 'master' of https://github.com/arnog/mathquill

arnog committed:
Reference.html correctly display UTF-8 chars

arnog committed:
Fix remaining significant differences with mainline

The following control sequences are normalized to the control sequence
indicated after ->:
* \neq -> \ne (note that \neq is a LaTeX command, but we previously
normalized it to \ne)
* \land -> \wedge (note that \land is a LaTeX command, but we previously
normalized it to \wedge)
* \lor -> \vee (note that \lor is a LaTeX command, but we previously
normalized it to \vee)
* \smallsetminus -> \setminus
* \oslash and \Oslash -> \varnothing = U+2205 (although they should map
to U+00F8 and U+00D8 respectively to be consistent with LaTeX)
* \upsih and \Upsih -> \Upsilon
* \sigmav -> \varsigma
* \opencurlybrace -> \lbrace
* \closecurlybrace -> \rbrace
* \vert is now mapped to U+007c. U+2223 is the \divides binary operator
* \| now maps to \Vert for consistency with LaTeX

arnog commented:
The previous commit addressed the latest issues raised. After this commit, the collation looks like the attached PDF. Note that \| is indeed defined, but it is an alias to \Vert for consistency with LaTeX (and MathJax and KaTex).

[collation.pdf](https://github.com/mathquill/mathquill/files/224967/collation.pdf)

jwmerrill committed:
Merge pull request #606 from mathquill/fix.sin-squared

Fix sin^2(x), and sin^2 x as cursor moves through #visual

jwmerrill committed:
Merge pull request #604 from mathquill/fix.autocmds-eating-op-names

Fix auto-commands eating letters from operator names

jwmerrill committed:
Merge pull request #603 from mathquill/feature.integrals

Feature: integrals' from/to now mandatory, prettier

stufflebear committed:
Editting of docs across the board

This includes spelling updates, clarifications, a couple new methods and larger rewrites of the Config and Contributing Guides

stufflebear committed:
Edits to Config and Contributing Guides

stufflebear committed:
Incorporate new README content

From commit: https://github.com/mathquill/mathquill/commit/92bb0eb49678331182fa80f4bd5d3685cfd7f406

stufflebear committed:
Fixed LaTeX return and added documentation for \\MathQuillMathField

Thanks @kschaefe for the PR comments

laughinghan committed:
Rename .fastClick() => .clickAt()

'clickAt' is a better name than 'fastClick' because:
- the signature is self-explanatory, whereas .fastClick() only makes
  sense to people who've read about FastClick/"Fast Buttons" [FN]
- this functionality is useful for more than just touch taps, I think
  that combined with .registerEmbed() (#539) this obviates
  .dropEmbedded() (#533).

[FN]
FastClick: https://github.com/ftlabs/fastclick
"Fast Buttons": https://web.archive.org/web/20150406091222/https://developers.google.com/mobile/articles/fast_buttons
(Google seems to have lost the original widely-cited "Fast Buttons" article: https://developers.google.com/mobile/articles/fast_buttons )

laughinghan committed:
Make `target` an optional argument to .clickAt()

It previously was the first argument but was unused, which, why.

It's optional because it's not fundamentally necessary to the
functionality, but it's nice to allow it as a micro-optimization:
it's one less DOM FFI call, and it'll be available in pretty much any
use case I can think of (touch tap, drag-and-drop a thing into
MathQuill).

laughinghan committed:
Re-org and comment "fast touch taps" test case

laughinghan committed:
Adapt .dropEmbedded() test for .cickAt()

Also fix .clickAt() and .ignoreNextMousedown() to allow chaining

laughinghan committed:
.clickAt() focuses the editable field

Also, still .preventDefault() even if .ignoreNextMousedown()-ing,
otherwise 300ms after a touch tap when the legacy mouse events fire on
the container span (because the container span has a 'mousedown' event
handler [1]), "clicking" on the container span, it would blur the
textarea (just like a real click would unless you .preventDefault()).

[1]:

> A clickable element is a link, form element, image map area, or any
> other element with mousemove, mousedown, mouseup, or onclick handlers.

https://developer.apple.com/library/ios/documentation/AppleApplications/Reference/SafariWebContent/HandlingEvents/HandlingEvents.html

laughinghan committed:
Move clickAt/ignoreNextMousedown to EditableField

laughinghan committed:
Merge pull request #622 from mathquill/feature.fastClick-rebase

API to enable "fast clicks" on touch devices

Argghhhhh. Want touch swipes to scroll the page, even if they start in
MathQuill (so that it's possible to scroll an expression list filled to
the brim with nothing but MathQuills), but touch taps to place the
cursor. Now, with mouse events, if all you care about is a logical
"click" on the current platform, you just listen for the click event and
the browser will take care of whether a particular sequence of
mousedown-mousemoves-mouseup is a click or not. With touch events,
there's no equivalent logical "tap" event, and while legacy mouse events
do fire if you don't .preventDefault() on touchstart, Safari on iOS waits
a glacial ~360ms [1] after a tap in case you're actually doing the
double-tap gesture to zoom in.

[1]: Videos: http://developer.telerik.com/featured/300-ms-click-delay-ios-8/

It gets better. Not only are the legacy mouse events unhelpful, they're
misleading, because there is no way to directly tell them apart from
real mouse events and no way to get them not to fire except calling
.preventDefault(), which breaks touch-swipe-to-scroll.

As a result, across the Web there are dozens of blog posts about, and
shitty JS libraries for, "fast clicks", that all disagree on what counts
as a "tap", and also when to ignore a mouse event because it's probably
a legacy one 300ms after a touch event that's already been dealt with.
By far the 2 most-cited "fast click" solutions are Google Developer's
"Fast Buttons" article [2] and FT Labs' FastClick library [3]; only the
latter ignores taps that are to stop scrolling after a fling, taps where
touchend is >700ms after touchstart, or taps with more than one finger,
but only the former appears to put a time and distance limit on mouse
events to ignore (FastClick just ignores the next one).

[2]: https://web.archive.org/web/20150406091222/https://developers.google.com/mobile/articles/fast_buttons
[3]: https://github.com/ftlabs/fastclick

(Google seems to have lost the original widely-cited "Fast Buttons" article: https://developers.google.com/mobile/articles/fast_buttons )

MathQuill's sponsor, Desmos, has a custom touch events library whose
definition of "tap", and whose strategy for ignoring delayed mouse
events, is different from either of the above. Clearly, this stuff needs
to be decided by the parent application using MathQuill.

So, with heart-wrenching despair, here are 2 new API calls:

- .clickAt(clientX, clientY, target) takes in the touch event info
  and places the cursor like there was a click there (also useful for
  e.g. obviating .dropEmbedded())
    + `target` is optional but optimizes-away one DOM call for free

- .ignoreNextMousedown(fn) takes a function that is called on subsequent
  mousedown events, which MathQuill will ignore if true is returned.
  MathQuill stops calling the function once it returns false.
  You should call this whenever there might be a legacy mousedown, even
  if it doesn't "count" as a tap to you, else MathQuill will get clicked
  fast sometimes and slow others, that'd be weird.

--

Why clientX/Y even though .seek() takes in pageX/Y? Well,
document.elementFromPoint() *has* to take in clientX/Y, whereas .seek()
merely *happens* to take in pageX/Y because it's slightly easier to work
with jQuery::offset(), and that's likely to change because jQuery::offset()
has poor performance so MathQuill may want to start working with
.getBoundingClientRects() directly, which would make it actually easier to
work with clientX/Y than pageX/Y in the first place.

For now, though, I'm not gonna port everything over the whole .seek() system
to clientX/Y yet, instead .clickAt() just converts clientX/Y to pageX/Y by
way of window.pageXOffset and window.pageYOffset, which according to
QuirksMode [4] is the difference between them and has very good support.
(It's also used by jQuery to convert from clientX/Y to what jQuery::offset()
returns [5].)

[4]: http://www.quirksmode.org/mobile/tableViewport.html#t10
[5]: https://github.com/jquery/jquery/blob/1.12.3/src/offset.js#L113-L114

--

The test case has a trivial notion of logical "tap", which is if no
touchmoves happen between the touchstart and touchend. Note that it
calls .ignoreNextMousedown() after every touchend, or else sometimes if
you wiggle your finger a little but not much, the logical "tap" won't
happen but Safari on iOS will still fire a legacy mousedown event.

Test case also gained -webkit-tap-highlight-color to to hide the gray
tap highlight box.

laughinghan committed:
Fix merge of visual test

https://github.com/mathquill/mathquill/commit/3de91d8f5a2f96f3a805dfd341010fcb4bba3ca9#commitcomment-17218984

dominique-unruh commented:
Currently (i.e., in c41aef4), macros such as \lVert and \rVert do not create brackets. The intended use in TeX is as brackets. (Otherwise, we'd just use \Vert.) I could not find a way to enter them (neither via text entry, not via LaTeX) as brackets.

laughinghan committed:
CircleCI + Sauce Labs

See comments inline in the code

laughinghan committed:
Fix circle.yml syntax error

In YAML square brackets can mean array syntax (like JSON), so gotta
heredoc that shit

laughinghan committed:
Attempt to optimize waiting on Sauce Connect

Cache[1] the binary rather than re-downloading and untarring it every
time, and if it isn't cached, download/untar/run it in the background in
parallel to the other shit (though the other shit is pretty much just
`npm install`).

[1]: https://circleci.com/docs/how-cache-works/
also https://circleci.com/docs/configuration/#cache-directories

laughinghan committed:
Ignore errors trying to cleanup Sauce Connect

Seems to be the only reason build#7 failed:
https://circleci.com/gh/mathquill/mathquill/7

laughinghan committed:
Fix circle.yml syntax

laughinghan committed:
Fix Pjs version

We use `Options.p` in a lot of places, but `.p` on Pjs classes wasn't
introduced until v3.1.0:
https://github.com/jneen/pjs/commit/0ce7b701d241fe7c0d8d12a172a2f353300adc7c

Thanks @robinwit!
https://mathquill.slack.com/archives/mathquill/p1461679514000080
https://mathquill.slack.com/archives/mathquill/p1461755912000101

jwmerrill committed:
Merge pull request #627 from mathquill/fix.visual-rendering

Improve visual rendering of integrals

jwmerrill committed:
PJs 5.0.0+ is currently incompatible with MathQuill.

PJs 5.0.0 introduced a new behavior for this.constructor(...) in order
to be compatible with coffeescript classes. The new behavior means
that, if you don't use `new` when you call `this.constructor`, the
object this is an alias of `this` instead of a new object.

This behavior interacts badly with MathQuill's use of PJs to subclass
jQuery, and jQuery's use of `this.constructor` inside `pushStack`:

https://github.com/jquery/jquery/blob/305f193aa57014dc7d8fa0739a3fefd47166cd44/src/core.js#L72-L84

laughinghan committed:
Merge pull request #629 from mathquill/pin-pjs-dep

Pin pjs to less >=3.1.0 <5.0.0

stufflebear committed:
Merge pull request #543 from Learnosity/feature/dollar-sign

[LRN] Allow direct entry of $ (dollar sign)

stufflebear committed:
Merge pull request #551 from mathquill/fix.stealing-back-focus-test

Upgrade to Mocha 2.4 to fix iffy "stealing back" focus test

stufflebear committed:
Merge pull request #580 from mathquill/build.nodejs-vs-node

Makefile nice err msg if no Node or installed wrong

stufflebear committed:
Added Docs information to README and docs/index

laughinghan committed:
Fix .dropEmbedded()/.clickAt() test instability

They were passing in Chrome 48, they're still passing in Firefox 45, but
they're failing in Chrome 50, missing the 'm' by 1 pixel. Now it's
solidly in the 'm' in all browsers.

laughinghan committed:
Fix #632 supSubsRequireOperand detaches selection

...but leaves it in the DOM

jwmerrill committed:
Merge pull request #634 from mathquill/fix.supSubsRequireOperand-selection

Fix #632 supSubsRequireOperand detaches selection

laughinghan committed:
Remove .html() from Getting Started docs

laughinghan committed:
Tweaks to Getting Started docs

laughinghan committed:
Fix typo in Getting Started docs

laughinghan committed:
Getting Started docs uses relative links not absolute

http://www.mkdocs.org/user-guide/writing-your-docs/#linking-documents

This way docs.mathquill.com vs mathquill.readthedocs.com is internally
consistent (though I wish mathquill.readthedocs.com just redirected to
docs.mathquill.com), and it's internally consistent locally without any
hosted docs in the first place.

laughinghan committed:
Undocument .html() method

laughinghan committed:
Revise wording of API Methods docs

Also fix all links to be relative

laughinghan committed:
Fix typo in API Methods docs

laughinghan committed:
Edits to Config docs like relative links and typos

laughinghan committed:
Fix links in Contributing docs to be relative links

laughinghan committed:
Symlink instead of duplicating README.md and docs/index.md

laughinghan committed:
Try README.md symlink instead of index.md symlink

It just copies it over, maybe it'll respect the symlink where Nginx or
mkdocs doesn't:
https://github.com/mkdocs/mkdocs/issues/321#issuecomment-70099532

laughinghan committed:
README.md symlinks to docs/README.md

Swap of previous commit's attempt

If there's one thing we learned from the previous commit, though
ReadTheDocs/MkDocs doesn't respect symlinks, GitHub does, so this should
be fine.

laughinghan committed:
Symlink all the things

laughinghan committed:
Oops forgot to `git add` new file docs/index.md

laughinghan committed:
Tweak README quickstart example

laughinghan committed:
Tiny tweak to README quickstart example

laughinghan committed:
<big> is ignored by both ReadTheDocs and GitHub >:(

laughinghan committed:
Favor Slack over Issues and IRC

laughinghan committed:
Tweak link to Slack in README

laughinghan committed:
Tweak intro wording to README

laughinghan committed:
Formatting tweaks to Code of Conduct

link and use @ for our handles everywhere

Also link coc@mathquill.com email, and change word "any maintainer" to
"any moderator"

laughinghan committed:
Fix README link to homepage

It's actually broken since for whatever reason it's redirecting to
HTTPS://mathquill.com, and GitHub Pages only supports HTTP not HTTPS for
custom domains.

laughinghan committed:
Move Slack badge to the README byline

laughinghan committed:
Add GIF of homepage demo to README

laughinghan committed:
Move thesis sentence above GIF in README

laughinghan committed:
Homepage demo GIF on README should link to homepage

stufflebear committed:
Merge pull request #623 from mathquill/docs.readthedocs

Refactor documentation from README into Read the Docs

desa committed:
Verify that something works

I'm going to go away as a commit

laughinghan committed:
Shim Mocha.process.stdout.write() inst of console.log()

Turns out Mocha deliberately exposes it:
https://github.com/mochajs/mocha/pull/916

laughinghan committed:
Elucidate Node server for CI to POST xunit.xml to

- Format the Node one-liner across multiple lines in a heredoc
- Write directly to $CIRCLE_TEST_REPORTS/ rather than later `cat` and
  redirecting
- Comment on why this is necessary

laughinghan committed:
Fix unit test page to still use HTML reporter by default

However, if you supply a query parameter `post_xunit_to` and URL, which
our CI now does, then it will POST the xUnit XML to the given URL.

Also, cleanup commented-out code.

Recommend -w to view diff.

laughinghan committed:
Fix xunit.xml test report for CI

Oops I forgot that this ends up in the XML file

laughinghan committed:
Merge pull request #625 from mathquill/build.circle-ci

[Build System] CircleCI + Sauce Labs

laughinghan committed:
Fix #534 "Better error message if jQuery isn't loaded"

controversial committed:
Merge pull request #637 from mathquill/fix.better-missing-jQ-err-msg

Fix #534 "Better error message if jQuery isn't loaded"

arnog committed:
Merge branch 'master' into mathquill/master

laughinghan commented:
Working on the merge conflicts with `master`, jQuery 1.5.2 is causing MathJax to behave funnily where the rendered math is all smooshed together at the end of the section instead of inside the table where it should be (switching back to jQuery 1.7.2 relieves the problem):

<img width="743" alt="screen shot 2017-02-21 at 18 26 02" src="https://cloud.githubusercontent.com/assets/225809/23194959/18ea0934-f866-11e6-90de-0964aa9c559f.png">

I tried generating wrapping `$$ ... $$`s but with just one call to `MathJax.Hub.Typeset()` (unlike what we'd tried before which called that many, many times: https://github.com/arnog/mathquill/commit/eb3299bd013d7aaadcd9b02d85b94a7fbb65a4c2#diff-76ea3ebad150f64578dbeb98a5cfcfc1L1291 ), and timed it compared to using jQuery 1.7.2, they both take ~9s, the `<script type="math/tex">` thing doesn't really save us any time, so I'm reverting us to that.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mathquill/mathquill?pullRequest=615) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/mathquill/mathquill?pullRequest=615) before we can accept your contribution.<br/>**4** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: laughinghan<br/>:white_check_mark: stufflebear<br/>:white_check_mark: controversial<br/>:white_check_mark: arnog<br/>:x: desa<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mathquill/mathquill?pullRequest=615) it.</sub>

520qzx520 commented:
mathquill How do I add latex symbols that mathquill does not support
